### PR TITLE
[Enhancement] Move io coalesce input stream below cache input stream (zero-copy feature)

### PR DESCRIPTION
**Pull request by dirtysalt**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

The reason why to put logic of io coalesce under cache input stream is because, io coalesce will amplify IO size which will affect block cache finally.  

For example, right now it's like  `[IOCoalesceStream] -> [CacheInputStream]`
- if we wants to read [10, 20], [30, 40]
- IOCoalesceStream will merge them into [10,40]
- CacheInputStream will cache [10,40]
- which is unecessary, because we don't need to cache [20,30]

So if we work like this  `[CacheInputStream] -> [IOCoalesceStream]`,  cache size won't be affect by io coalesce.

----

However the downside of this is that we can not use following API any more
```
    virtual Status get_bytes(const uint8_t** buffer, size_t offset, size_t* nbytes, bool peek) = 0;
```
So there will be an extra memcpy.

__UPDATED:__  since I've implemented zero copy read bytes, times of extra memcpy is less
- if data is from block cache, because block cache memory is segmented, we have to use memcpy to gather them into a single piece
- but if data is from shared buffer stream, we can use shared buffer directly without copy.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Comment by dirtysalt** [comment]
run starrocks_admit_test

File: be/src/util/buffered_stream.cpp [hunk 1]
```diff
@@ -16,77 +16,14 @@
...
-
-Status DefaultBufferedInputStream::get_bytes(const uint8_t** buffer, size_t offset, size_t* nbytes, bool peek) {
-    seek_to(offset);
-    return get_bytes(buffer, nbytes, peek);
-}
-
-// ===================================================================================
-
-SharedBufferedInputStream::SharedBufferedInputStream(RandomAccessFile* file) : _file(file) {}
+SharedBufferedInputStream::SharedBufferedInputStream(std::shared_ptr<SeekableInputStream> stream) : _stream(stream) {}
```
Comment (Youngwb) [inline be/src/util/buffered_stream.cpp:26]: Why not move these to be/src/io/shared_buffered_input_stream.cpp?

**Comment by Youngwb** [comment]
I have a question, for example

1. if we wants to read [10, 20]
2. CacheInputStream already cached [10,15}, [16,20]
3. will use IOCoalesceStream to read [15] ?
4. if IOCoalesceStream's IO range is [15-20], it means stiil need to read [16-20] from parquet/orc even if it already cached?

File: be/src/util/buffered_stream.cpp (hunk 1 above)
Comment (dirtysalt) [reply be/src/util/buffered_stream.cpp:26]: because all  input stream classes are supposed to be in `io` namespace`

**Comment by dirtysalt** [comment]
> I have a question, for example
> 
> 1. if we wants to read [10, 20]
> 2. CacheInputStream already cached [10,15}, [16,20]
> 3. will use IOCoalesceStream to read [15] ?
> 4. if IOCoalesceStream's IO range is [15-20], it means stiil need to read [16-20] from parquet/orc even if it already cached?


If `CacheInputStream` alreay cache [15], there is no request hit on IOCoalesceStream at all. `CacheInputStream` will intercept all requests if data is cached.

**Comment by dirtysalt** [comment]
run starrocks_admit_test

**Comment by dirtysalt** [comment]
run starrocks_admit_test

**Review by Youngwb — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Comment by dirtysalt** [comment]
And to compare IO counter reduction, we have to change code for a little bit for branch-3.0: to move `CountedInputStream` to lowest level.

```
diff --git a/be/src/exec/hdfs_scanner.cpp b/be/src/exec/hdfs_scanner.cpp
index 2d5c4bd..1aaf57c 100644
--- a/be/src/exec/hdfs_scanner.cpp
+++ b/be/src/exec/hdfs_scanner.cpp
@@ -209,6 +209,10 @@ Status HdfsScanner::open_random_access_file() {
         input_stream = std::make_shared<io::CompressedSeekableInputStream>(compressed_input_stream);
     }

+    // input_stream = CountedInputStream(input_stream)
+    // NOTE: make sure `CountedInputStream` is last applied, so io time can be accurately timed.
+    input_stream = std::make_shared<CountedSeekableInputStream>(input_stream, &_stats);
+
     // if block cache
     // input_stream = CacheInputStream(input_stream)
     if (_scanner_params.use_block_cache && _compression_type == CompressionTypePB::NO_COMPRESSION) {
@@ -217,10 +221,6 @@ Status HdfsScanner::open_random_access_file() {
         input_stream = _cache_input_stream;
     }

-    // input_stream = CountedInputStream(input_stream)
-    // NOTE: make sure `CountedInputStream` is last applied, so io time can be accurately timed.
-    input_stream = std::make_shared<CountedSeekableInputStream>(input_stream, &_stats);
-
     // so wrap function is f(x) = (CountedInputStream (CacheInputStream (DecompressInputStream x)))
     _file = std::make_unique<RandomAccessFile>(input_stream, _raw_file->filename());
     _file->set_size(_scanner_params.file_size);
```

And io counter comparison is following.



  | main io | 3.0 io
-- | -- | --
Q01 | 4852 | 9085
Q02 | 4852 | 9085
Q03 | 4852 | 9085
Q04 | 6019 | 10421
Q05 | 6020 | 10009
Q06 | 6020 | 10009
Q07 | 6023 | 10280
Q08 | 6023 | 11037
Q09 | 6023 | 10226
Q10 | 5498 | 8399
Q11 | 6904 | 11997
Q12 | 6904 | 12328
Q13 | 6019 | 12264

**Comment by dirtysalt** [comment]
I've made a very dedicated experiment to test effect of this PR.

the SQL is following: we select out max two fields.

> select max(lo_orderkey), max(lo_linenumber) from lineorder_flat;


And in parquet file, between two fields there is one field whose size is about 1MB, which means two fields can be read by coalesced way.

```

row group 8:      RC:648773 TS:108250505 OFFSET:870542356
--------------------------------------------------------------------------------
lo_orderkey:       INT32 UNCOMPRESSED DO:0 FPO:870542356 SZ:2595237/2595237/1.00 VC:648773 ENC:BIT_PACKED,PLAIN,RLE
lo_orderdate:      INT32 UNCOMPRESSED DO:0 FPO:873137593 SZ:984263/984263/1.00 VC:648773 ENC:BIT_PACKED,PLAIN_DICTIONARY,RLE
lo_linenumber:     INT32 UNCOMPRESSED DO:0 FPO:874121856 SZ:244770/244770/1.00 VC:648773 ENC:BIT_PACKED,PLAIN_DICTIONARY,RLE
lo_custkey:        INT32 UNCOMPRESSED DO:0 FPO:874366626 SZ:2595237/2595237/1.00 VC:648773 ENC:BIT_PACKED,PLAIN,RLE
lo_partkey:        INT32 UNCOMPRESSED DO:0 FPO:876961863 SZ:2595237/2595237/1.00 VC:648773 ENC:BIT_PACKED,PLAIN,RLE
lo_suppkey:        INT32 UNCOMPRESSED DO:0 FPO:879557100 SZ:2595237/2595237/1.00 VC:648773 ENC:BIT_PACKED,PLAIN,RLE
lo_orderpriority:  BINARY UNCOMPRESSED DO:0 FPO:882152337 SZ:245051/245051/1.00 VC:648773 ENC:BIT_PACKED,PLAIN_DICTIONARY,RLE
```

----

before PR, block cache read 2.56GB

```
- BlockCache: 0
  - BlockCacheReadBytes: 2.56 GB
    - __MAX_OF_BlockCacheReadBytes: 4.90 MB
    - __MIN_OF_BlockCacheReadBytes: 16.00 KB
  - BlockCacheReadCounter: 6.626K (6626)
    - __MAX_OF_BlockCacheReadCounter: 12
    - __MIN_OF_BlockCacheReadCounter: 1
  - BlockCacheReadTimer: 6.476ms
    - __MAX_OF_BlockCacheReadTimer: 62.906ms
    - __MIN_OF_BlockCacheReadTimer: 25.747us
  - BlockCacheWriteBytes: 0.00 
  - BlockCacheWriteCounter: 0
  - BlockCacheWriteFailBytes: 0.00 
  - BlockCacheWriteFailCounter: 0
  - BlockCacheWriteTimer: 0ns
- RequestBytesRead: 2.47 GB
  - __MAX_OF_RequestBytesRead: 4.43 MB
  - __MIN_OF_RequestBytesRead: 14.36 KB
```

and with this PR, block cache read 2.49GB

```

- BlockCache: 0
  - BlockCacheReadBytes: 2.49 GB
    - __MAX_OF_BlockCacheReadBytes: 4.33 MB
    - __MIN_OF_BlockCacheReadBytes: 16.00 KB
  - BlockCacheReadCounter: 16.612K (16612)
    - __MAX_OF_BlockCacheReadCounter: 33
    - __MIN_OF_BlockCacheReadCounter: 1
  - BlockCacheReadTimer: 6.62ms
    - __MAX_OF_BlockCacheReadTimer: 30.650ms
    - __MIN_OF_BlockCacheReadTimer: 30.422us
  - BlockCacheWriteBytes: 0.00 
  - BlockCacheWriteCounter: 0
  - BlockCacheWriteFailBytes: 0.00 
  - BlockCacheWriteFailCounter: 0
  - BlockCacheWriteTimer: 0ns
- RequestBytesRead: 2.47 GB
- __MAX_OF_RequestBytesRead: 4.30 MB
- __MIN_OF_RequestBytesRead: 14.33 KB

```

**Comment by dirtysalt** [comment]
I used the following case to get it to work

```
-- Q03. for tpch-100g, there are 574 rows.
select * from lineitem where l_suppkey = 768951;
```

I added a delayed materialization case
- RequestBytesRead = 6.51GB
- before BlockCacheReadBytes = 34.39 GB
- after BlockCacheReadBytes = 6.56GB

Saving 20% of the previous value

**Comment by dirtysalt** [comment]
@mergify backport branch-3.0

**Pull request merged**